datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -272,6 +273,13 @@ impl MemoryConsumer for ExternalSorter {
     }
 
     async fn spill(&self) -> Result<usize> {
+        let partition = self.partition_id();
+        let mut in_mem_batches = self.in_mem_batches.lock().await;
+        // we could always get a chance to free some memory as long as we are holding some
+        if in_mem_batches.len() == 0 {
+            return Ok(0);
+        }
+
         debug!(
```

Comment:
I just moved the debug message down as it was confusing that spill didn't actually happen if `in_mem_batches` was empty